File: benchmarks/binarytrees/ruby/trace_bench.rb
# The Computer Language Benchmarks Game
# http://shootout.alioth.debian.org
#
# contributed by Jesse Millikan
# Modified by Wesley Moxam and Michael Klaus


def item_check(left, item, right)
  puts "def item_check(left, item, right)"
  if left.nil? then
    puts "if left.nil? then"
    return item
  end
  item + item_check(*left) - item_check(*right)
end

def bottom_up_tree(item, depth)
  puts "def bottom_up_tree(item, depth)"
  if depth == 0 then
    puts "if depth == 0 then"
    return [nil, item, nil] 
  end
  item_item = 2 * item
  depth -= 1
  [bottom_up_tree(item_item - 1, depth), item, bottom_up_tree(item_item, depth)]
end

def run_iter(max_depth)
    puts "def run_iter(max_depth)"
    #max_depth = ARGV[0].to_i
    min_depth = 4

    max_depth = [min_depth + 2, max_depth].max

    stretch_depth = max_depth + 1
    stretch_tree = bottom_up_tree(0, stretch_depth)

    #puts "stretch tree of depth #{stretch_depth}\t check: #{item_check(*stretch_tree)}"
    stretch_tree = nil

    long_lived_tree = bottom_up_tree(0, max_depth)

    base_depth = max_depth + min_depth
    min_depth.step(max_depth + 1, 2) do |depth|
      puts "min_depth.step(max_depth + 1, 2) do |depth|"
      iterations = 2 ** (base_depth - depth)

      check = 0

      for i in 1..iterations
        puts "for i in 1..iterations"
        temp_tree = bottom_up_tree(i, depth)
        check += item_check(*temp_tree)

        temp_tree = bottom_up_tree(-i, depth)
        check += item_check(*temp_tree)
      end

      #puts "#{iterations * 2}\t trees of depth #{depth}\t check: #{check}"
    end

    #puts "long lived tree of depth #{max_depth}\t check: #{item_check(*long_lived_tree)}"
end

File: benchmarks/binarytrees/php/trace_bench.php
<?php 
/* The Computer Language Benchmarks Game
   http://benchmarksgame.alioth.debian.org/

   contributed by Peter Baltruschat
   modified by Levi Cameron
   modified by Craig Russell
 */

class Tree {
   public $i;
   public $l;
   public $r;
   
   public function __construct($item, $depth) {
      echo 'public function __construct($item, $depth) {', "\n";
      $this->i = $item;
      if($depth) {
         echo 'if($depth) {', "\n";
         $this->l = new Tree($item * 2 - 1, --$depth);
         $this->r = new Tree($item * 2, $depth);
      }
   }
   
   public function check() {
      echo 'public function check() {', "\n";
      return $this->i
         + ($this->l->l === null ? $this->l->check() : $this->l->i)
         - ($this->r->l === null ? $this->r->check() : $this->r->i);
   }
}


function run_iter($n) {
    echo 'function run_iter($n) {', "\n";
    $minDepth = 4;

    //$n = $argc == 2 ? $argv[1] : 1;
    $maxDepth = $minDepth + 2 > $n ? $minDepth + 2 : $n;
    $stretchDepth = $maxDepth + 1;

    $stretch = new Tree(0, $stretchDepth);
    //printf("stretch tree of depth %d\t check: %d\n",
    //   $stretchDepth, $stretch->check());
    unset($stretch);

    $longLivedTree = new Tree(0, $maxDepth);

    $iterations = 1 << $maxDepth;
    do {
       echo 'do {', "\n";
       $check = 0;
       for($i = 1; $i <= $iterations; ++$i) {
          echo 'for($i = 1; $i <= $iterations; ++$i) {', "\n";
          $check += (new Tree($i, $minDepth))->check()
             + (new Tree(-$i, $minDepth))->check();
       }

       //printf("%d\t trees of depth %d\t check: %d\n",
       //   $iterations<<1, $minDepth, $check);

       $minDepth += 2;
       $iterations >>= 2;
    }
    while($minDepth <= $maxDepth);

    //printf("long lived tree of depth %d\t check: %d\n",
    //   $maxDepth, $longLivedTree->check());
}
run_iter(1)
?>
